azkaban-db/src/main/java/azkaban/db/MySQLDataSource.java

```diff
@@ -58,18 +65,126 @@ public synchronized Connection getConnection() throws SQLException {
          * Every Attempt generates a thread-hanging-time, about 75 seconds, which is hard coded, and can not be changed.
          */
         connection = createDataSource().getConnection();
-        if (connection != null) {
+
+        /**
+         * If connection is null or connection is read only, retry to find available connection.
+         * When DB fails over from master to slave, master is set to read-only mode. We must keep
+         * finding correct data source and sql connection.
+         */
+        if (connection == null || isReadOnly(connection)) {
+          throw new SQLException("Failed to find DB connection Or connection is read only. ");
+        } else {
           return connection;
         }
       } catch (final SQLException ex) {
-        logger.error(
-            "Failed to find DB connection. waits 1 minutes and retry. No.Attempt = " + retryAttempt,
-            ex);
-      } finally {
+
+        logger.error( "Failed to find write-enabled DB connection. Wait 1 minutes and retry."
+            + " No.Attempt = " + retryAttempt, ex);
```

Comment:
should it be (retryAttempt+1) to match with No.Attempt?

Reply:
retry attempt starts from 0. Isn't correct?

Reply:
It's a bit nit. But if log shows No.attempt = 1, i think generally people will assume it's first attempt.

Reply:
Good point.